image_writers_guide/guidelines.adoc

```diff
@@ -9,4 +9,85 @@
+
+=== Use Volumes for Persisted Data
+Images should expect to use a https://docs.docker.com/reference/builder/#volume[Docker volume] for persisted data.  This way the OpenShift system can mount network storage to the node where the container is running and then reattach that storage at a new node if the container moves.  By using the volume for all persisted storage needs, the content will be preserved throughout container restarts and moves.  If your image writes data to arbitrary locations within the container, there is no guarantee that content will be preserved.
+
+By explicitly defining volumes in your Dockerfile, you make it easy for consumers of the image to understand what volumes they need to define when running your image.
+
+For more information on how Volumes are used in OpenShift, see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/volumes.md[this documentation]. (NOTE to docs team:  this link should really go to something in the openshift docs, once we have it)
+
+Keep in mind that even with persisted volumes, each instance of your image will have its own volume, the filesystem is not shared between instances.  This means the volume cannot be used to share state in a cluster.
+
+=== Use Services for inter-image communication
+For cases where your image needs to communicate with a service provided by another image (e.g. a web frontend image that needs to access a database image to store/retrieve data), your image should consume an OpenShift Service.  Services provide a static endpoint for access which will not change as containers are stopped/started/moved.  In addition services provide load balancing for requests.  For more information see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/services.md[this documentation].  (NOTE to docs team:  this link should really go to something in the openshift docs once we have it)
+
+=== Provide Common Libraries
+Ensure that your image contains commonly used libraries for your platform.  In particular, provide database drivers for common databases used with your platform (e.g. provide JDBC drivers for MySQL and PostgreSQL if you are creating a java framework image).  This will prevent the need for common dependencies to be downloaded during application assembly time, speeding up application image builds.  It will also simplify the work application authors need to do to ensure all their dependencies are met.
+
+=== Configuration
+Runtime configuration should be handled via environment variables.  This means that configuration files should be parameterized with environment variables and the values should be substituted in during image startup.  This allows consumers of your image to customize behavior (e.g. database settings, passwords, performance tuning) without having to introduce a new layer on top of your image.  Instead, they can simply define environment variable values when defining a pod, and change those settings without rebuilding the image.
```

Comment:
Would say "Simple runtime configuration is most easily handled in Docker with environment variables.  The more complex your configuration becomes, you may want to create wrapper scripts that parameterize configuration files.  For very complex configurations, where environment variables are totally unsuitable, expect to receive configuration files via volumes.  If you need to do this, be sure to provide clear error messages on image start to end users." as the first few sentences.

Also, highlight that secrets like certs and keys can be passed as environment variables (like passwords).  It's often better to ensure no secrets ever *live* in the image.